proposals/params-span.md

```diff
@@ -0,0 +1,146 @@
+# `params Span<T>` and implicit allocation of arrays
+
+## Summary
+Avoid heap allocation for implicit allocation of arrays in specific scenarios such as `params` arguments.
+
+## Motivation
+`params` array parameters provide a convenient way to call a method that takes an arbitrary length list of arguments.
+However, using an array type for the parameter means the compiler must implicitly allocate an array on the heap at each call site.
+
+If we extend `params` types to include the `ref struct` types `Span<T>` and `ReadOnlySpan<T>`, where values of those types cannot escape the call stack, the array at the call site may be created on the stack instead.
+
+And if we're extending `params` to other types, we  could also allow `params IEnumerable<T>` to avoid allocating and copying collections at call sites that have an `IEnumerable<T>` rather than `T[]`.
+
+The benefits of `params ReadOnlySpan<T>` and `params Span<T>` are primarily for new APIs. Existing commonly used APIs such as `Console.WriteLine()` and `StringBuilder.AppendFormat()` already have overloads that avoid array allocations for common cases.
+```csharp
+public static class Console
+{
+    public static void WriteLine(string value);
+    public static void WriteLine(string format, object arg0);
+    public static void WriteLine(string format, object arg0, object arg1);
+    public static void WriteLine(string format, object arg0, object arg1, object arg2);
+    public static void WriteLine(string format, params object[] arg);
+}
+```
+
+## Detailed design
+
+### Extending `params`
+`params` parameters will be supported with types `Span<T>`, `ReadOnlySpan<T>`, and `IEnumerable<T>`.
+
+A call in _expanded_ form to a method with a `params T[]` or `params IEnumerable<T>` parameter will result in an array `T[]` allocated on the heap.
```

Comment:
>_expanded_ form [](http://example.com/codeflow?start=10&length=15)

1. Consider including a link to spec section that defines "expanded form" [https://github.com/dotnet/csharplang/blob/main/spec/expressions.md#applicable-function-member](https://github.com/dotnet/csharplang/blob/main/spec/expressions.md#applicable-function-member)
2. Is allocating a `T[]` array on the heap the only option for `params IEnumerable<T>` in expanded form?

Reply:
I've added a link for _expanded form_, thanks.

For `params IEnumerable<T>`, the collection will need to be allocated on the heap since the method being called may hold onto a reference to the collection. Presumably it could be an `IEnumerable<T>` other than `T[]` though.